core-to-plc/src/Language/Plutus/CoreToPLC.hs

```diff
@@ -391,19 +510,22 @@ getDataCons tc' = sortConstructors tc' <$> extractDcs tc'
 -- See Note [Scott encoding of datatypes]
 convDataCons :: forall m. Converting m => GHC.TyCon -> [GHC.DataCon] -> m PLCType
 convDataCons tc dcs = do
-    abstracted <- flip (foldr (\tv acc -> mkTyLam tv acc)) (GHC.tyConTyVars tc) $ do
-        resultType <- safeFreshTyName $ (GHC.getOccString $ GHC.getName tc) ++ "_matchOut"
-        cases <- mapM (dataConCaseType (PLC.TyVar () resultType)) dcs
-        pure $ mkScottTyBody resultType cases
+    abstracted <- flip (foldr (\tv acc -> mkTyLam tv acc)) (GHC.tyConTyVars tc) $ mkScottTy tc dcs
 
-    tds <- gets csTypeDefs
+    tcDefs <- gets csTypeDefs
     isSelfRecursive <- isSelfRecursiveTyCon tc
     if isSelfRecursive
-    then case Map.lookup (GHC.getName tc) tds of
-        Just (InProgress tyname) -> pure $ PLC.TyFix () tyname abstracted
-        _                        -> throwPlain $ ConversionError "Could not find var to fix over"
+    then case Map.lookup (GHC.getName tc) tcDefs of
+        Just (Def{dVis=Abstract,dVar=(tyname, _)}, _) -> pure $ PLC.TyFix () tyname abstracted
```

Comment:
So before definitions are constructed, they're abstract, then they're visible and then they're abstract again? That's a bit confusing. Or am I reading this wrongly?

Reply:
Yes, that's right. If you read the note carefully, you'll see that in the first stage we bind them to the variable that we will fix over. This is a bit I'm not totally happy with, but we represent that as an abstract definition of a type name.

Reply:
I used to have a separate kind of thing for this, but it was very noisy to work with. I'm still not very happy with the way this bit in particular works (finding which variable to fix over).